From aa4a7ee1d3417bc11207ad0a78d579ece7fe0c13 Mon Sep 17 00:00:00 2001
From: Felix Fietkau <nbd@nbd.name>
Date: Fri, 17 Oct 2025 10:25:55 +0000
Subject: [PATCH] ubusd: fix more instances of missing length checks for
 patterns

More instances of the same bug fixed by commit d31effb4277bd5.

Reported-by: Karsten Sperling <ksperling@apple.com>
Signed-off-by: Felix Fietkau <nbd@nbd.name>
(cherry picked from commit aa4a7ee1d3417bc11207ad0a78d579ece7fe0c13)
[Backport to ubus 2022-02-21-b32a0e17 (OpenWrt 19.07, Gluon v2021.1.x) by
 Freifunk im Neanderland <projekt@neanderfunk.de>: code unchanged,
 hunks refreshed for line offsets only. Part of the CVE-2025-62526 fix
 set (d31effb, aa4a7ee, 60e0404), as in OpenWrt 24.10 a790196993.]
---
 ubusd_acl.c   | 17 +++++++++++++++++
 ubusd_proto.c |  3 +++
 2 files changed, 20 insertions(+)

diff --git a/ubusd_acl.c b/ubusd_acl.c
--- a/ubusd_acl.c
+++ b/ubusd_acl.c
@@ -236,6 +236,9 @@ ubusd_acl_alloc_obj(struct ubusd_acl_file *file, const char *obj)
 	char *k;
 	bool partial = false;
 
+	if (!len)
+		return NULL;
+
 	if (obj[len - 1] == '*') {
 		partial = true;
 		len--;
@@ -266,6 +269,8 @@ ubusd_acl_add_access(struct ubusd_acl_file *file, struct blob_attr *obj)
 		return;
 
 	o = ubusd_acl_alloc_obj(file, blobmsg_name(obj));
+	if (!o)
+		return;
 
 	o->methods = tb[ACL_ACCESS_METHODS];
 	o->tags = tb[ACL_ACCESS_TAGS];
@@ -280,6 +285,9 @@ ubusd_acl_add_subscribe(struct ubusd_acl_file *file, const char *obj)
 {
 	struct ubusd_acl_obj *o = ubusd_acl_alloc_obj(file, obj);
 
+	if (!o)
+		return;
+
 	o->subscribe = true;
 }
 
@@ -288,6 +296,9 @@ ubusd_acl_add_publish(struct ubusd_acl_file *file, const char *obj)
 {
 	struct ubusd_acl_obj *o = ubusd_acl_alloc_obj(file, obj);
 
+	if (!o)
+		return;
+
 	o->publish = true;
 }
 
@@ -295,6 +306,9 @@ static void ubusd_acl_add_listen(struct ubusd_acl_file *file, const char *obj)
 {
 	struct ubusd_acl_obj *o = ubusd_acl_alloc_obj(file, obj);
 
+	if (!o)
+		return;
+
 	o->listen = true;
 }
 
@@ -302,6 +316,9 @@ static void ubusd_acl_add_send(struct ubusd_acl_file *file, const char *obj)
 {
 	struct ubusd_acl_obj *o = ubusd_acl_alloc_obj(file, obj);
 
+	if (!o)
+		return;
+
 	o->send = true;
 }
 
diff --git a/ubusd_proto.c b/ubusd_proto.c
--- a/ubusd_proto.c
+++ b/ubusd_proto.c
@@ -201,6 +201,9 @@ static int ubusd_handle_lookup(struct ubus_client *cl, struct ubus_msg_buf *ub,
 
 	objpath = blob_data(attr[UBUS_ATTR_OBJPATH]);
 	len = strlen(objpath);
+	if (!len)
+		return UBUS_STATUS_INVALID_ARGUMENT;
+
 	if (objpath[len - 1] != '*') {
 		obj = avl_find_element(&path, objpath, obj, path);
 		if (!obj)
